File: 2026MOB/PAM/test_NPQ.m
data = readtable('test_NPQ.RPT', FileType='text')

colors   = {'#1f77b4', '#2ca02c', '#ff7f0e', '#8B0000'};
markers  = {'x', 'd', 'x', 'o'};
mfcolors = colors;

labels = { ...
    '\lambda_1 = 470 nm', ...
    '\lambda_2 = 520 nm', ...
    '\lambda_3 = 645 nm', ...
    '\lambda_4 = 665 nm'};

mfcolors{4} = 'none';

% Conversion temps HH:MM:SS -> secondes
sec_ = @(s) sscanf(s, '%d:%d:%d')' * [3600; 60; 1];

% Temps complet
time_all = zeros(height(data),1);
for i = 1:height(data)
    time_all(i) = sec_(char(data.Time(i)));
end

figs_NPQ = gobjects(4,1);

keys = {'AllOpePSII', 'opePSII', 'maxPSII', 'NPQEstim'};

for i = 1:length(keys)

    k = keys{i};

    figs_NPQ(i) = figure;
    ax = axes;
    hold(ax, 'on');

    for ch = 1:4

        c   = colors{ch};
        m   = markers{ch};
        mfc = mfcolors{ch};
        lbl = labels{ch};

        chs = num2str(ch);

        time = time_all;
        time0 = time(1);

        % Colonnes dynamiques
        Fm = data.(sprintf('Fm%s', chs));
        F  = data.(sprintf('F%s', chs));

        if strcmp(k, 'AllOpePSII')
            Y = (Fm - F) ./ Fm;
        end

        % Données sous lumière actinique
        idx_AL = data.PAR > 16;
        data_AL = data(idx_AL,:);

        if ~strcmp(k, 'AllOpePSII')

            time = zeros(height(data_AL),1);

            for t = 1:height(data_AL)
                time(t) = sec_(char(data_AL.Time(t)));
            end

            fm = data_AL.(sprintf('Fm%s', chs));
            f  = data_AL.(sprintf('F%s', chs));

            fq = fm - f;

            if strcmp(k, 'opePSII')
                Y = fq ./ fm;
            end

            % Ligne suivante après chaque mesure AL
            idx = find(idx_AL);
            idx_next = idx + 1;
            idx_next(idx_next > height(data)) = [];

            data_0 = data(idx_next,:);

            f0_star  = data_0.(sprintf('F%s', chs));
            f0m_star = data_0.(sprintf('Fm%s', chs));

            % Adapter longueur si nécessaire
            n = min([length(f0_star), length(fm)]);

            f0_star  = f0_star(1:n);
            f0m_star = f0m_star(1:n);
            fm2      = fm(1:n);

            f0 = 1 ./ (1 ./ f0_star - 1 ./ f0m_star + 1 ./ fm2);

            fv = fm2 - f0;

            if strcmp(k, 'maxPSII')
                Y = fv ./ fm2;
            end

            % Données dark adapted
            idx_first_AL = find(idx_AL,1,'first');
            data_DA = data(1:idx_first_AL-1,:);

            fm_DA = data_DA.(sprintf('Fm%s', chs));
            fm_DA = fm_DA(end);

            if strcmp(k, 'NPQEstim')
                Y = (fm_DA - fm2) ./ fm2;
            end
        end

        % Valeurs négatives -> NaN
        Y(Y < 0) = NaN;

        plot((time - time0)/60, Y, ...
            'LineStyle', 'none', ...
            'Marker', m, ...
            'Color', c, ...
            'MarkerFaceColor', mfc, ...
            'DisplayName', lbl);
    end

    xlabel('Time / min');

    if strcmp(k, 'AllOpePSII')
        ylabel('Yield');
    end

    if strcmp(k, 'opePSII')
        ylabel({'PSII operating efficiency', ...
            'K_P''[Q_A]'' / (K_F + K_{NPQ}'' + K_P''[Q_A]'')'});
    end

    if strcmp(k, 'maxPSII')
        ylabel({'PSII maximal efficiency', ...
            'K_P'' / (K_F + K_{NPQ}'' + K_P'')'});
    end

    if strcmp(k, 'NPQEstim')
        ylabel('NPQ: (K_{NPQ}'' - K_{NPQ}) / K_{NPQ}');
    end

    legend('show');

    % Axe secondaire PAR
    ax.Box = 'off';
    yyaxis right

    time2 = repelem(time_all, 2);
    time2 = [time2(1)-20 ; time2];

    PAR = repelem(data.PAR, 2);

    plot((time2(1:end-1) - time2(1))/60, PAR, ...
        'Color', [0.5 0.5 0.5], ...
        'LineWidth', 1, ...
        'HandleVisibility', 'off');

    ylabel('PAR / \muE');
    ax.YColor = [0.5 0.5 0.5];

    yyaxis left
end

File: 2026MOB/PAM/test_NPQ.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
mpl.rcParams['axes.labelsize'] = 20
mpl.rcParams['axes.titlesize'] = 20
mpl.rcParams['xtick.direction'] = 'in'
mpl.rcParams['xtick.major.size'] = 10
mpl.rcParams['xtick.major.width'] = 2.5
mpl.rcParams['xtick.minor.size'] = 8
mpl.rcParams['xtick.minor.width'] = 1.2
mpl.rcParams['xtick.labelsize'] = 15
mpl.rcParams['xtick.top'] = True
mpl.rcParams['ytick.direction'] = 'in'
mpl.rcParams['ytick.major.size'] = 14
mpl.rcParams['ytick.major.width'] = 2.5
mpl.rcParams['ytick.minor.size'] = 8
mpl.rcParams['ytick.minor.width'] = 1.2
mpl.rcParams['ytick.labelsize'] = 15
mpl.rcParams['ytick.right'] = True
mpl.rcParams['legend.fontsize'] = 15
mpl.rcParams['axes.linewidth'] = 2.5
mpl.rcParams['lines.linewidth'] = 3
mpl.rcParams['lines.markersize'] = 10

data = pd.read_csv('test_NPQ.RPT', sep='\s+')

colors = ['C0', 'C2', 'C1', 'darkred']
markers = ['x', 'D', 'x', 'o']
mfcolors = [c for c in colors]
labels = [r'$\lambda_1=470\,$nm', r'$\lambda_2=520\,$nm', r'$\lambda_3=645\,$nm', r'$\lambda_4=665\,$nm']
mfcolors[3] = 'None'

def sec_(s):
    l = s.split(':')
    return int(l[0]) * 3600 + int(l[1]) * 60 + int(l[2])
sec = np.vectorize(sec_)

figs_NPQ = {}
for i, k in enumerate(['AllOpePSII', 'opePSII', 'maxPSII', 'NPQEstim']):
    figs_NPQ[i], ax = plt.subplots()
    for ch, (c, m, mfc, lbl) in enumerate(zip(colors, markers, mfcolors, labels), start=1):
        ch = str(ch)
        time = sec(data['Time'])
        time0 = time[0]
        if k == 'AllOpePSII': Y = (data['Fm'+ch]-data['F'+ch])/data['Fm'+ch]

        data_AL = data[data['PAR']>16]
        if k != 'AllOpePSII': time = sec(data_AL['Time'])
        fm = np.asarray(data_AL['Fm'+ch])
        f = np.asarray(data_AL['F'+ch])
        fq = fm - f
        if k == 'opePSII': Y = fq/fm  
            
        data_0 = data.iloc[data_AL.index+1]
        f0_star = np.asarray(data_0['F'+ch])
        f0m_star = np.asarray(data_0['Fm'+ch])
        f0 = 1/(1/f0_star-1/f0m_star+1/fm)
        fv = fm-f0 
        if k == 'maxPSII': Y = fv/fm

        data_DA = data.iloc[:data_AL.index[0]]
        fm_DA = data_DA['Fm'+ch].iloc[-1]
        if k == 'NPQEstim': Y = (fm_DA-fm)/fm
        
        Y[Y<0] = np.nan
        plt.plot((time - time0)/60, Y, lw=0, marker=m, color=c, markerfacecolor=mfc, label = lbl)
        # plt.plot((time - time[0])/60, Y*data['PAR']*0.84*0.5, lw=0, marker=m, color=c, markerfacecolor=mfc)
        
    # ax.set_ylim([0, None])
    # ax.set_xlim([-10, 130])
    ax.set_xlabel('Time  /  min')
    if k == 'AllOpePSII': ax.set_ylabel('Yield')
    if k == 'opePSII': ax.set_ylabel('PSII operating efficiency\n'
                                     +r"$K_\mathrm{P}'[Q_A]'/(K_\mathrm{F}+K_\mathrm{NPQ}'+K_\mathrm{P}'[Q_A]')$")
    if k == 'maxPSII': ax.set_ylabel('PSII maximal efficiency\n'
                                     +r"$K_\mathrm{P}'/(K_\mathrm{F}+K_\mathrm{NPQ}'+K_\mathrm{P}')$")     
    if k == 'NPQEstim': ax.set_ylabel('NPQ: ' + r"$(K_\mathrm{NPQ}'-K_\mathrm{NPQ})/K_\mathrm{NPQ}$")
    plt.legend()

    ax.spines["right"].set_visible(False)
    ax_AL = ax.twinx()
    time = sec(data['Time'])
    time = np.repeat(time, 2)
    time = [time[0]-20]+list(time)
    PAR = list(np.repeat(data['PAR'], 2))
    ax_AL.plot((time[:-1] - time[0])/60, PAR, alpha=0.35, color = 'grey')
    ax_AL.set_ylabel('PAR  /  µE')
    ax_AL.yaxis.label.set_color('grey')
    ax_AL.spines["right"].set_edgecolor('grey')
    ax_AL.tick_params(axis='y', colors='grey')